compositor_render/src/scene/transition.rs

```diff
@@ -0,0 +1,96 @@
+use std::time::Duration;
+
+use super::{types::interpolation::InterpolationState, InterpolationKind};
+
+/// Similar concept to InterpolationState, but it represents a time instead.
+/// Values between 0 and 1 represent transition and larger than 1 post transition.
+///
+/// If interpolation_kind is linear then InterpolationState and TransitionProgress
+/// have the same numerical value.
+#[derive(Debug, Clone, Copy)]
+struct TransitionProgress(f64);
+
+#[derive(Debug, Clone)]
+pub(super) struct TransitionState {
+    /// Additional offset for transition. It is non zero if you want to start
+    /// a transition in the middle of the interpolation curve.
+    initial_offset: (TransitionProgress, InterpolationState),
+
+    // PTS of a first frame of transition.
+    start_pts: Duration,
+
+    /// Duration of the transition.
+    duration: Duration,
+
+    interpolation_kind: InterpolationKind,
+}
+
+pub(super) struct TransitionOptions {
+    pub duration: Duration,
+    pub interpolation_kind: InterpolationKind,
+}
+
+impl TransitionState {
+    pub fn from_previous_transition(
+        current_transition: Option<TransitionOptions>,
+        previous_transition: Option<TransitionState>,
+        last_pts: Duration,
+    ) -> Option<Self> {
+        let previous_transition = previous_transition.and_then(|transition| {
+            if transition.start_pts + transition.duration < last_pts {
+                return None;
+            }
+            Some(transition)
+        });
+        match (current_transition, previous_transition) {
+            (None, None) => None,
+            (None, Some(previous_transition)) => {
+                let remaining_duration = (previous_transition.start_pts
+                    + previous_transition.duration)
+                    .saturating_sub(last_pts);
+                let progress_offset = TransitionProgress(
+                    1.0 - (remaining_duration.as_secs_f64()
+                        / previous_transition.duration.as_secs_f64()),
+                );
```

Comment:
This only makes sense for the same transition type*. Is this expected?

* I can explain it in more detail in person. Long story short, if transition functions have different values at transition progress (like one is linear, the other is cubic), this would cause discontinuity in the interpolation function (e.g. rapid change of corners rounding radius).

Reply:
This is a case `(None, Some(previous_transition))` so scene update where previous transition is still in progress and new update does not have a transition defined yet. So we are only dealing here with one type of transition.

And whether it will work for non linear transition: I think it works, I tested it with some basic examples, but not enough to be sure.